lib/improver/optical_flow/optical_flow.py

```diff
@@ -0,0 +1,301 @@
+
+        cond1 = point_in_bounds(xsrc_point_frac, ysrc_point_frac, xdim, ydim)
+        adv_field[cond1] = 0
+
+        # Find the integer points surrounding the fractional source coordinates
+        # and check they are in bounds
+        xsrc_point_lower = xsrc_point_frac.astype(int)
+        xsrc_point_upper = xsrc_point_lower + 1
+        ysrc_point_lower = ysrc_point_frac.astype(int)
+        ysrc_point_upper = ysrc_point_lower + 1
+
+        cond2 = point_in_bounds(xsrc_point_lower, ysrc_point_lower,
+                                xdim, ydim) & cond1
+        cond3 = point_in_bounds(xsrc_point_lower, ysrc_point_upper,
+                                xdim, ydim) & cond1
+        cond4 = point_in_bounds(xsrc_point_upper, ysrc_point_lower,
+                                xdim, ydim) & cond1
+        cond5 = point_in_bounds(xsrc_point_upper, ysrc_point_upper,
+                                xdim, ydim) & cond1
+
+        # Calculate the distance-weighted fractional contribution of points
+        # surrounding the source coordinates
+        x_weight_upper = xsrc_point_frac - xsrc_point_lower.astype(float)
+        x_weight_lower = 1. - x_weight_upper
+        y_weight_upper = ysrc_point_frac - ysrc_point_lower.astype(float)
+        y_weight_lower = 1. - y_weight_upper
+
+        # Advect data from each of the four source points onto the output grid
+        self._increment_output_array(data, adv_field, cond2, xgrid, ygrid,
+                                     xsrc_point_lower, ysrc_point_lower,
+                                     x_weight_lower, y_weight_lower)
+        self._increment_output_array(data, adv_field, cond3, xgrid, ygrid,
+                                     xsrc_point_upper, ysrc_point_lower,
+                                     x_weight_upper, y_weight_lower)
+        self._increment_output_array(data, adv_field, cond4, xgrid, ygrid,
+                                     xsrc_point_lower, ysrc_point_upper,
+                                     x_weight_lower, y_weight_upper)
+        self._increment_output_array(data, adv_field, cond5, xgrid, ygrid,
+                                     xsrc_point_upper, ysrc_point_upper,
+                                     x_weight_upper, y_weight_upper)
+
+        return adv_field
+
+    def process(self, cube, timestep, bgd=0.0):
+        """
+        Extrapolates input cube data and updates validity time.  The input
+        cube should have precisely two non-scalar dimension coordinates
+        (spatial x/y), and is expected to be in a projection such that grid
+        spacing is the same (or very close) at all points within the spatial
+        domain.  The input cube should also have a "time" coordinate.
+
+        Args:
+            cube (iris.cube.Cube):
+                The 2D cube containing data to be advected
+            timestep (datetime.timedelta):
+                Advection time step
+            bgd (float):
```

Comment:
My thinking here was that the default value is a property of the cube diagnostic to be advected rather than the advection velocity fields.  Is there a good reason to move it?

Reply:
The main reason for moving it was that as part of `self`, it needn't be passed around all the methods and would thus reduce the number of arguments and comments relating to it.

Reply:
Just saving repetition.
 If it were likely that we were to setup one instance of AdvectField, passing different cubes with different backgrounds through that plugin, I see the concern.